src/transformers/modeling_fsmt.py

```diff
@@ -0,0 +1,893 @@
+# coding=utf-8
+# Copyright 2020 The Facebook AI Research Team Authors and The HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Original implementation: https://github.com/pytorch/fairseq/tree/master/examples/wmt19
+# Authors:
+# - @alexeib Alexei Baevski
+# - @edunov Sergey Edunov
+# - @michaelauli Michael Auli
+# - @myleott Myle Ott
+# - @nng555 Nathan Ng
+# - David Grangier
+# - Kyra Yee
+#
+# Paper: Facebook FAIR's WMT19 News Translation Task Submission https://arxiv.org/abs/1907.06616
+#
+"""PyTorch Fairseq model, ported from https://github.com/pytorch/fairseq/tree/master/examples/wmt19"""
+
+import logging
+import math
+import random
+import warnings
+from typing import Any, List, Optional, Tuple
+
+import torch
+import torch.nn.functional as F
+from torch import Tensor, nn
+from torch.nn import CrossEntropyLoss
+
+from .configuration_fsmt import FSMTConfig
+from .file_utils import (
+    add_code_sample_docstrings,
+    add_end_docstrings,
+    add_start_docstrings,
+    add_start_docstrings_to_callable,
+    replace_return_docstrings,
+)
+from .modeling_bart import (
+    DecoderLayer,
+    EncoderLayer,
+    LayerNorm,
+    LearnedPositionalEmbedding,
+    _prepare_bart_decoder_inputs,
+    _reorder_buffer,
+    invert_mask,
```

Comment:
Before merging, could we try to have as little as possible from other modeling files? `EncoderLayer` and `DecoderLayer` are only 50 lines each, copy-pasting them in this file would simplify users' work when studying the model, as they won't have to switch between files.

For reference, this is one of the main takeaways from the survey that you can fine [here](https://discuss.huggingface.co/t/transformers-huge-community-feedback/120), with "easy-to-tweak, self-contained" as likes and "Too much modularization - Model code is spread across too many files" as dislikes.

Reply:
Well, I had them duplicated copied first and then it was suggested I remove them and import from Bart instead. So I did. I am not sure how to proceed.

Reply:
A simplified model that isn't dependent on another model's parts or code would definitely be the way to go here, this is what we reach for with other models in the library. Let's wait for @sshleifer to chime in.